plugin/storage/es/factory_test.go

```diff
@@ -238,4 +237,48 @@ func TestInitFromOptions(t *testing.T) {
 	f.InitFromOptions(o)
 	assert.Equal(t, o.GetPrimary(), f.primaryConfig)
 	assert.Equal(t, o.Get(archiveNamespace), f.archiveConfig)
+}*/
+
+func TestPasswordFromFile(t *testing.T) {
+	passwordFile, err := os.CreateTemp("", "")
+	require.NoError(t, err)
+	defer passwordFile.Close()
+
+	passwordFile.WriteString("bar")
+
+	c := escfg.Configuration{
+		Username:         "foo",
+		PasswordFilePath: passwordFile.Name(),
+	}
+
+	f := NewFactory()
+	f.newClientFn = func(c *escfg.Configuration, logger *zap.Logger, metricsFactory metrics.Factory) (es.Client, error) {
+		rawClient := &elastic.Client{}
+
+		passwordFromFile, err := escfg.LoadFileContent(c.PasswordFilePath)
+		require.NoError(t, err)
+		c.Password = passwordFromFile
+
+		option := elastic.SetBasicAuth(c.Username, c.Password)
+		option(rawClient)
+
+		return eswrapper.WrapESClient(rawClient, nil, 0), nil
+	}
+	f.primaryConfig.Store(&c)
+	f.archiveConfig.Store(&c)
+
+	require.NoError(t, f.Initialize(metrics.NullFactory, zap.NewNop()))
+	assert.Equal(t, "bar", f.primaryConfig.Load().Password)
+
+	primaryClient, err := f.newClientFn(f.primaryConfig.Load(), nil, nil)
+	require.NoError(t, err)
+
+	var expectedPrimaryClient atomic.Pointer[es.Client]
+	expectedPrimaryClient.Store(&primaryClient)
+	assert.Equal(t, expectedPrimaryClient.Load(), f.primaryClient.Load())
+
+	_, err = passwordFile.WriteString("baz")
+	require.NoError(t, err)
+	f.onPrimaryPasswordChange()
```

Comment:
I don't know why but I have to explicitly call f.onPrimaryPasswordChange() here to make the password change. Don't know if it's a problem of the watcher in the implementation or a problem in my test.

Reply:
Also, because watcher is running in a separate goroutine, you cannot immediately validate if the client was updated, usually we use assert.Eventually (grep for examples in this repo)